src/java/org/apache/cassandra/streaming/StreamManager.java

```diff
@@ -260,10 +269,20 @@ public StreamManager()
         logger.info("Storing streaming state for {} or for {} elements", duration, numElements);
         states = CacheBuilder.newBuilder()
                              .expireAfterWrite(duration.quantity(), duration.unit())
-                             .maximumSize(numElements)
+                             .maximumWeight(10 * 1024 *1024)
+                             .weigher(new streamingStateWeigher())
                              .build();
     }
 
+    private class streamingStateWeigher implements Weigher<TimeUUID,StreamingState>
+    {
+        public int weigh(TimeUUID key, StreamingState val) {
+            long weight = ObjectSizes.measureDeep(val);
```

Comment:
performance issue, this is a costly operation which is why we don't use it outside of static access.  You should look at the "empty" size we already computed and compute the "change" from empty.

Reply:
to make this cleaner, you prob want to implement `IMeasurableMemory` in StreamingState and just have this method call it, you can then compute the logic there.